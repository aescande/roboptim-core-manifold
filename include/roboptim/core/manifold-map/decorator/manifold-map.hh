// Copyright (C) 2015 by Grégoire Duchemin, AIST, CNRS, EPITA
//                       Félix Darricau, AIST, CNRS, EPITA
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_MANIFOLD_MAP_HH
# define ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_MANIFOLD_MAP_HH

# include <roboptim/core/manifold-map/decorator/descriptive-wrapper.hh>
# include <roboptim/core/manifold-map/decorator/manifold-desc.hh>
# include <roboptim/core/manifold-map/decorator/function-on-manifold.hh>

#include <manifolds/SO3.h>
#include <manifolds/RealSpace.h>
#include <manifolds/CartesianProduct.h>
#include <manifolds/ExpMapMatrix.h>
#include <manifolds/S2.h>

#define ROBOPTIM_DESCRIPTIVE_FWD_DECS(NAME) typedef NAME type

#define REAL_SPACE(num) roboptim::Real<num>::Space
#define DESC_MANIFOLD(name, ...) typedef roboptim::ManiDesc< __VA_ARGS__> name
#define DEFINE_MANIFOLD_FROM_FUNCTION(name) template<class U>\
  struct Manifold_##name{\
  static pgs::Manifold* getInstance(U* function);\
  };\
  template<class U>\
  pgs::Manifold* Manifold_##name <U>::getInstance(U* function)
#define DEFINE_MANIFOLD(name) template<class U>\
  struct Manifold_##name{\
  static pgs::Manifold* getInstance(U* function);\
  };\
  template<class U>\
  pgs::Manifold* Manifold_##name <U>::getInstance(U*)

#define BIND_FUNCTION_ON_MANIFOLD(function, manifold) typedef roboptim::DescriptiveWrapper<function, manifold> function##_On_##manifold; \
  typedef roboptim::FunctionOnManifold<typename function::parent_t> Instance_##function##_On_##manifold
#define NAMED_FUNCTION_BINDING(name, function, manifold) typedef roboptim::DescriptiveWrapper<function, manifold> name; \
  typedef roboptim::FunctionOnManifold<typename function::parent_t> Instance_##name

// Library-defined elementary descriptive manifolds
// I do not think we should put those in a namespace of their own,
// because the user has to use the classes' names when (s)he defines
// a manifold.
//
// FIXME: define all remaining manifolds

namespace roboptim {

  template <class FI>
  struct S2
  {
    static pgs::Manifold* getInstance(FI*)
    {
      return new pgs::S2();
    }
  };

  template <class FI>
  struct SO3
  {
    static pgs::Manifold* getInstance(FI*)
    {
      return new pgs::SO3<pgs::ExpMapMatrix>();
    }
  };

  template<int I = 1>
  struct Real
  {
    template <class FI>
    struct Space
    {
      static pgs::Manifold* getInstance(FI*)
      {
	return new pgs::RealSpace(I);
      }
    };
  };

  template<class FI>
  struct Automated_Real
  {
  public:
    static pgs::Manifold* getInstance(FI* function)
    {
      return new pgs::RealSpace(function->getSize());
    }
  };

}

#endif //! ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_MANIFOLD_MAP_HH
