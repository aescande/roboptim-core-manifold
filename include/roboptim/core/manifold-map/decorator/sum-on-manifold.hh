// Copyright (C) 2015 by Félix Darricau, AIST, CNRS, EPITA
//                       Grégoire Duchemin, AIST, CNRS, EPITA
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_SUM_ON_MANIFOLD_HH
# define ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_SUM_ON_MANIFOLD_HH

# include <vector>
# include <iostream>
# include <utility>
# include <type_traits>
# include <memory>
# include <functional>

# include <roboptim/core/detail/autopromote.hh>
# include <roboptim/core/manifold-map/decorator/descriptive-wrapper.hh>
# include <roboptim/core/manifold-map/decorator/function-on-manifold.hh>
# include <roboptim/core/manifold-map/decorator/manifold-merger.hh>

# include <manifolds/Manifold.h>

# include <boost/noncopyable.hpp>

namespace roboptim
{
  /// \addtogroup roboptim_manifolds
  /// @{

  template<typename T>
  class AdderOnManifold;

  /// \brief This RobOptim function computes a weighted sum of
  /// FunctionOnManifold. It is defined on a manifold which
  /// merges all elementary manifolds of the functions it sums
  /// as merged by a ManifoldMerger.
  ///
  /// Since messing up the arguments and wrapping around it is
  /// easy, it can only be constructed through an appropriate
  /// factory, AdderOnManifold.
  ///
  /// \tparam T RobOptim Eigen matrix type
  template<typename T>
  class SumOnManifold
    : public GenericTwiceDifferentiableFunction<T>
  {
    ROBOPTIM_DEFINE_FLAG_TYPE();
  public:
    ROBOPTIM_TWICE_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_ (GenericTwiceDifferentiableFunction<T>);

    typedef FunctionOnManifold<T> function_t;
    typedef std::shared_ptr<function_t> functionPtr_t;

    /// This function can only be created through a factory
    friend class AdderOnManifold<T>;

  private:
    /// \brief shared pointers to the functions to sum
    std::vector<functionPtr_t> functions_;
    /// \brief weight of each function in the sum
    std::vector<double> weights_;

    /// \brief the type flags of this function sum
    flag_t flags_;

    /// \brief a buffer for the result of computing this sum
    mutable result_t resultBuffer_;
    /// \brief a buffer for the result of computing the gradient of this sum
    mutable gradient_t gradientBuffer_;
    /// \brief a buffer for the result of computing the jacobian of this sum
    mutable jacobian_t jacobianBuffer_;
    /// \brief a buffer for the result of computing the hessian of this sum
    mutable hessian_t hessianBuffer_;

    /// \brief constructs a SumOnManifold
    ///
    /// \param functions a vector of functions to be summed
    /// \param weights a vector of weights, one for each function
    /// \param inputSize the inputSize of this sum (representationDim of the merged manifold)
    /// \param outputSize the outputSize of all summed functions
    /// \param name the name of this sum
    SumOnManifold(std::vector<functionPtr_t>& functions,
		  std::vector<double> weights,
		  size_type inputSize, size_type outputSize, std::string name);

  public:
    void impl_compute (result_ref result, const_argument_ref x)
      const;

    void impl_gradient (gradient_ref,
                        const_argument_ref,
                        size_type)
      const;

    void impl_jacobian (jacobian_ref jacobian,
                        const_argument_ref x)
      const;

    void impl_hessian(hessian_ref hessian,
		      const_argument_ref x,
		      size_type functionId = 0)
      const;

    virtual flag_t getFlags() const;
  };

  /// \brief This class is a factory used to create a weighted
  /// sum of FunctionOnManifold functions.
  ///
  /// \tparam T RobOptim Eigen matrix type
  template<typename T>
  class AdderOnManifold
  {
  public:
    typedef FunctionOnManifold<T> function_t;
    typedef std::shared_ptr<function_t> functionPtr_t;

  private:
    typedef typename std::function<functionPtr_t (const mnf::Manifold&)>
      descWrap_storage_t;

    /// \brief DescriptiveWrappers of functions to be summed.
    /// Stored as lambda functions to instantiate the FunctionOnManifold
    /// upon receiving the merged manifold.
    std::vector<descWrap_storage_t> functionsToSum_;
    /// \brief weights of each function
    std::vector<double> weights_;
    /// \brief utility member to compute the merged manifold of all summed functions
    ManifoldMerger merger_;

  public:
    /// \brief Default constructor
    AdderOnManifold(){}

    /// \brief Adds a function to the sum
    ///
    /// \tparam U type of the function being added
    /// \tparam V ManiDesc of the DescriptiveWrapper being added. Irrelevant here.
    ///
    /// \param weight weight fo the function in the weighted sum
    /// \param descWrap a DescriptiveWrapper<U, V> representing the function to be added to the sum
    /// \param instanceManifold the manifold on which the function will be instantiated
    /// \param restricted list of restricted manifolds
    /// \param restrictions list of restrictions for the restricted manifolds
    template<typename U, typename V>
    void add(double weight, DescriptiveWrapper<U, V>& descWrap, const mnf::Manifold& instanceManifold, std::vector<const mnf::Manifold*>& restricted, std::vector<std::pair<long, long>>& restrictions);
    template<typename U, typename V>
    void add(DescriptiveWrapper<U, V>& descWrap, const mnf::Manifold& instanceManifold, std::vector<const mnf::Manifold*>& restricted, std::vector<std::pair<long, long>>& restrictions);

    template<typename U, typename V>
    void add(DescriptiveWrapper<U, V>& descWrap, const mnf::Manifold& instanceManifold);

    template<typename U, typename V>
    void add(double weight, DescriptiveWrapper<U, V>& descWrap, const mnf::Manifold& instanceManifold);

    /// \brief Instantiate and return a shared pointer to a function computing the
    /// weighted sum of all functions added through the add(...) methods.
    ///
    /// \param globMani the manifold of the problem on which the function will be evaluated
    functionPtr_t getFunction(const mnf::Manifold& globMani) const;

    /// \brief clears out all fields of this instance
    void clear();

    /// \brief returns the merged manifold of all added functions
    const mnf::Manifold* getManifold() const;

    /// \brief return the number of functions added to the sum
    size_t numberOfFunctions();
  };

  /// @}

} // end of namespace roboptim.


# include <roboptim/core/manifold-map/decorator/sum-on-manifold.hxx>
#endif //! ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_SUM_ON_MANIFOLD_HH
