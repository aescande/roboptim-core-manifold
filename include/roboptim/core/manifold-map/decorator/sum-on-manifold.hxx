// Copyright (C) 2015 by Félix Darricau, AIST, CNRS, EPITA
//                       Grégoire Duchemin, AIST, CNRS, EPITA
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_SUM_ON_MANIFOLD_HXX
# define ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_SUM_ON_MANIFOLD_HXX

# include <roboptim/core/manifold-map/decorator/wrapper-on-manifold.hh>

namespace roboptim
{

  template<typename T>
  SumOnManifold<T>::SumOnManifold(std::vector<functionPtr_t>& functions,
		std::vector<double> weights,
		size_type inputSize, size_type outputSize, std::string name)
    : GenericTwiceDifferentiableFunction<T>(inputSize, outputSize, name),
      functions_(functions),
      weights_(weights),
      resultBuffer_(outputSize),
      gradientBuffer_(inputSize),
      jacobianBuffer_(outputSize, inputSize),
      hessianBuffer_(inputSize, inputSize)
  {
    flags_ = functions_[0]->getFlags();

    for (auto func : functions_)
      {
	flags_ &= func->getFlags();
      }
  }

  template<typename T>
  void SumOnManifold<T>::impl_compute (result_ref result, const_argument_ref x)
    const
  {
    result.setZero();
    size_t i = 0;
    for (auto function : functions_)
      {
	resultBuffer_.setZero();
	(*function)(resultBuffer_, x);
	result += weights_[i++] * resultBuffer_;
      }
  }

  template<typename T>
  void SumOnManifold<T>::impl_gradient (gradient_ref gradient,
		      const_argument_ref x,
		      size_type functionId)
    const
  {
    gradient.setZero();
    size_t i = 0;
    for (auto function : functions_)
      {
	gradientBuffer_.setZero();
	function->gradient(gradientBuffer_, x, functionId);
	gradient += weights_[i++] * gradientBuffer_;
      }
  }

  template<typename T>
  void SumOnManifold<T>::impl_jacobian (jacobian_ref jacobian,
		      const_argument_ref x)
    const
  {
    jacobian.setZero();
    size_t i = 0;
    for (auto function : functions_)
      {
	jacobianBuffer_.setZero();
	function->jacobian(jacobianBuffer_, x);
	jacobian += weights_[i++] * jacobianBuffer_;
      }
  }

  template<typename T>
  void SumOnManifold<T>::impl_hessian(hessian_ref hessian,
		    const_argument_ref x,
		    size_type functionId)
    const
  {
    hessian.setZero();
    size_t i = 0;
    for (auto function : functions_)
      {
	hessianBuffer_.setZero();
	function->hessian(hessianBuffer_, x, functionId);
	hessian += weights_[i++] * hessianBuffer_;
      }
  }

  template<typename T>
  typename SumOnManifold<T>::flag_t SumOnManifold<T>::getFlags() const
  {
    return flags_;
  }


  // ---- //

  template<typename T>
  template<typename U, typename V>
  void AdderOnManifold<T>::add(double weight, DescriptiveWrapper<U, V>& descWrap, mnf::Manifold& instanceManifold, std::vector<const mnf::Manifold*>& restricted, std::vector<std::pair<long, long>>& restrictions)
  {
    descWrap_storage_t lambda =
      [&descWrap, this, &instanceManifold, restricted, restrictions]
      (const mnf::Manifold& globMani)
      {
	return functionPtr_t
	(new WrapperOnManifold<T>
	 (descWrap, globMani, instanceManifold, restricted, restrictions)
	 );
      };

    functionsToSum_.push_back(lambda);
    weights_.push_back(weight);
    merger_.addManifold(instanceManifold);
  }

  template<typename T>
  template<typename U, typename V>
  void AdderOnManifold<T>::add(DescriptiveWrapper<U, V>& descWrap, mnf::Manifold& instanceManifold, std::vector<const mnf::Manifold*>& restricted, std::vector<std::pair<long, long>>& restrictions)
  {
    add(1.0, descWrap, instanceManifold, restricted, restrictions);
  }

  template<typename T>
  template<typename U, typename V>
  void AdderOnManifold<T>::add(DescriptiveWrapper<U, V>& descWrap, mnf::Manifold& instanceManifold)
  {
    std::vector<const mnf::Manifold*> restricted;
    std::vector<std::pair<long, long>> restrictions;

    add(descWrap, instanceManifold, restricted, restrictions);
  }

  template<typename T>
  template<typename U, typename V>
  void AdderOnManifold<T>::add(double weight, DescriptiveWrapper<U, V>& descWrap, mnf::Manifold& instanceManifold)
  {
    std::vector<const mnf::Manifold*> restricted;
    std::vector<std::pair<long, long>> restrictions;

    add(weight, descWrap, instanceManifold, restricted, restrictions);
  }

  template<typename T>
  typename AdderOnManifold<T>::functionPtr_t
  AdderOnManifold<T>::getFunction(const mnf::Manifold& globMani)
  {
    mnf::Manifold* sumManifold = merger_.getManifold();

    std::vector<functionPtr_t> functions;
    std::string name;

    for (descWrap_storage_t& lambda : functionsToSum_)
      {
	functions.push_back(lambda(*sumManifold));

	if (functions.size() > 1 &&
	    functions.back()->outputSize() != functions[functions.size() - 2]->outputSize())
	  {
	    std::cerr << "ERROR: function "<< functions.back()->getName()
		      << " of outputSize " << functions.back()->outputSize()
		      << " cannot be added to a sum of functions of outputSize " << functions[functions.size() - 2]->outputSize()
		      << std::endl;

	    throw std::runtime_error("Invalid size when summing FunctionOnManifold");
	  }

	name += (name.empty() ? "\"":", \"") + functions.back()->getName() + "\"";
      }
    name = "Sum(" + name + ")";

    typedef DescriptiveWrapper<SumOnManifold<T>, ManiDesc<>> descWrap_t;

    // FIXME: plug the leak
    SumOnManifold<T>* sumFunction = new SumOnManifold<T>(functions,
							 weights_,
							 functions[0]->inputSize(),
							 functions[0]->outputSize(),
							 name);

    descWrap_t* descWrap = descWrap_t::makeUNCHECKEDDescriptiveWrapper(sumFunction, *sumManifold);

    return functionPtr_t (new WrapperOnManifold<T> (*descWrap, globMani, *sumManifold));
  }

  template<typename T>
  void AdderOnManifold<T>::clear()
  {
    functionsToSum_.clear();
    weights_.clear();
    merger_.clear();
  }

  template<typename T>
  mnf::Manifold* AdderOnManifold<T>::getManifold()
  {
    return merger_.getManifold();
  }

  template<typename T>
  size_t AdderOnManifold<T>::numberOfFunctions()
  {
    return functionsToSum_.size();
  }

}

#endif //! ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_SUM_ON_MANIFOLD_HXX
