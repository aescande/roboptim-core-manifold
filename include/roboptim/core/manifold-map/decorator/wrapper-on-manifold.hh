// Copyright (C) 2015 by Félix Darricau, AIST, CNRS, EPITA
//                       Grégoire Duchemin, AIST, CNRS, EPITA
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_WRAPPER_ON_MANIFOLD_HH
# define ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_WRAPPER_ON_MANIFOLD_HH

# include <vector>
# include <iostream>
# include <utility>
# include <type_traits>

# include <roboptim/core/detail/autopromote.hh>
# include <roboptim/core/manifold-map/decorator/descriptive-wrapper.hh>
# include <roboptim/core/manifold-map/decorator/function-on-manifold.hh>

# include <manifolds/Manifold.h>

# include <boost/noncopyable.hpp>

namespace roboptim
{
  /// \addtogroup roboptim_manifolds
  /// @{

  /// \brief Maps a DescriptiveWrapper to a instance of a submanifold.
  ///
  /// \tparam T Matrix type
  template <typename T>
  class WrapperOnManifold :
    public FunctionOnManifold<T>
  {
    ROBOPTIM_DEFINE_FLAG_TYPE();
  public:
    ROBOPTIM_TWICE_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_ (GenericTwiceDifferentiableFunction<T>);

    /// \brief Creates the mapping between the manifolds.
    ///
    /// Programmers should ideally not directly use this constructor, amd
    /// rely on the macros defined in the file manifold-map.hh to get the well
    /// defined WrapperOnManifold type, ready to be instantiated.
    ///
    /// \tparam V input function type templating the DescriptiveWrapper.
    /// \tparam W descriptive manifold type templating the DescriptiveWrapper.
    ///
    /// \param descwrap instance of the DescriptiveWrapper
    /// \param problemManifold the manifold describing the whole variable vector.
    /// \param functionManifold the manifold describing the function's input vector.
    /// \param restrictedManifolds a list of elementary Manifolds to be restricted to a part of themselves
    /// \param restrictions the restrictions applying to the selected manifolds, represented as (startingIndex, size). If a single one is given, it will apply to all restricted manifolds.
    template <typename V, typename W>
    explicit WrapperOnManifold
    (DescriptiveWrapper<V, W>& descWrap,
     const mnf::Manifold& problemManifold,
     const mnf::Manifold& functionManifold,
     std::vector<const mnf::Manifold*> restrictedManifolds,
     std::vector<std::pair<long, long>> restrictions)
      :
      FunctionOnManifold<T>
      (static_cast<size_type>(problemManifold.representationDim()),
       descWrap.fct().outputSize (),
       (boost::format ("%1%")
	% descWrap.fct().getName ()).str ()),
      fct_ (&descWrap.fct()),
      fctDiff_ (0),
      fctTwiceDiff_ (0),
      manifold_ (&functionManifold),
      mappingFromFunction_ ()
    {
      computeMapping(descWrap,
                     problemManifold,
                     functionManifold,
                     restrictedManifolds,
                     restrictions);
      if (fct_->template asType<GenericDifferentiableFunction<T>>())
        fctDiff_ = fct_->template castInto<GenericDifferentiableFunction<T>>();
      this->mappedGradient_ = gradient_t(static_cast<int> (this->mappingFromFunction_.size()));
      this->mappedJacobian_ = jacobian_t(descWrap.fct().outputSize(), static_cast<int> (this->mappingFromFunction_.size()));
      this->tangentMappedJacobian = Eigen::MatrixXd::Zero(descWrap.fct().outputSize(), this->tangentMappingFromFunctionSize_);
      this->mappedGradient_.setZero();
      this->mappedJacobian_.setZero();
      if (fct_->template asType<GenericTwiceDifferentiableFunction<T>>())
        fctTwiceDiff_ = fct_->template castInto<GenericTwiceDifferentiableFunction<T>>();
      this->mappedHessian_ = hessian_t(descWrap.fct().inputSize(), static_cast<int> (this->mappingFromFunction_.size()));
      this->mappedHessian_.setZero();
    }


    /// \brief Creates the mapping between the manifolds without restrictions.
    ///
    /// Programmers should ideally not directly use this constructor, amd
    /// rely on the macros defined in the file manifold-map.hh to get the well
    /// defined WrapperOnManifold type, ready to be instantiated.
    ///
    /// \tparam V input function type templating the DescriptiveWrapper.
    /// \tparam W descriptive manifold type templating the DescriptiveWrapper.
    ///
    /// \param fct instance of the DescriptiveWrapper
    /// \param problemManifold the manifold describing the whole variable vector.
    /// \param functionManifold the manifold describing the function's input vector.
    template<typename V, typename W>
    explicit WrapperOnManifold (DescriptiveWrapper<V, W>& fct,
                                 const mnf::Manifold& problemManifold,
                                 const mnf::Manifold& functionManifold)
      : WrapperOnManifold(fct, problemManifold, functionManifold,
			   std::vector<const mnf::Manifold*>(), std::vector<std::pair<long, long>>())
    {
    }

    /// \brief WrapperOnManifold destructor
    ~WrapperOnManifold();

    /// \brief Traits type.
    typedef typename parent_t::traits_t traits_t;

    /// \brief Print method.
    /// \param o output stream.
    std::ostream& print(std::ostream& o) const;

    /// \brief apply the jacobian on the manifold's tangent space
    virtual void manifold_jacobian (mnf::RefMat jacobian,
				    const_argument_ref arg)
      const;

  protected:
    void impl_compute (result_ref result, const_argument_ref x)
      const;

    template <typename V, typename W>
    void computeMapping(DescriptiveWrapper<V, W>& descWrap,
			const mnf::Manifold& problemManifold,
			const mnf::Manifold& functionManifold,
			std::vector<const mnf::Manifold*> restrictedManifolds,
			std::vector<std::pair<long, long>> restrictions);

    /// \brief map the input to the restricted problem
    ///
    /// \param argument the argument to map
    void mapArgument(const_argument_ref argument)
      const;

    /// \brief the function.
    const GenericFunction<T>* fct_;

    /// \brief the function, differentiable access.
    const GenericDifferentiableFunction<T>* fctDiff_;

    /// \brief the function, twice differentiable access.
    const GenericTwiceDifferentiableFunction<T>* fctTwiceDiff_;
    /// \brief the problem manifold.
    const mnf::Manifold* manifold_;

    /// \brief vector representing the restricted mapping
    std::vector<size_t> mappingFromFunction_;

    /// \brief array representing the restricted mapping for the tangent problem
    size_t* tangentMappingFromFunction_;
    /// \brief size of the array
    long tangentMappingFromFunctionSize_;

    /// \brief new input mapped to the restricted problem
    mutable vector_t mappedInput_;

    void impl_gradient (gradient_ref gradient,
                        const_argument_ref argument,
                        size_type functionId = 0)
      const;
    void impl_jacobian (jacobian_ref jacobian,
                        const_argument_ref arg)
      const;

    /// \brief gets the gradient from the restricted problem
    ///
    /// \param gradient the output gradient
    void unmapGradient(gradient_ref) const;

    /// \brief unmap the jacobian from the restricted problem
    ///
    /// \param jacobian the jacobian to unmap
    void unmapTangentJacobian(mnf::RefMat jacobian)
      const;

    /// \brief new gradient mapped to the restricted problem
    mutable gradient_t mappedGradient_;
    /// \brief new jacobian mapped to the restricted problem
    mutable jacobian_t mappedJacobian_;

    // Dirty copy ONLY
    mutable Eigen::MatrixXd tangentMappedJacobian;

    void impl_hessian(hessian_ref, const_argument_ref, size_type) const;

    /// \brief gets the hessian from the restricted problem
    ///
    /// \param hessian the output hessian
    void unmapHessian(hessian_ref hessian) const;

    /// \brief new hessian mapped to the restricted problem
    mutable hessian_t mappedHessian_;

    /// \brief gets the jacobian from the restricted problem
    ///
    /// \param instance the WrapperOnManifold
    /// \param jacobian the output jacobian
    void unmapJacobian(jacobian_ref jacobian) const;

    /// \brief sets the jacobian on the manifold's tangent space with the
    /// computed value
    ///
    /// \param instance the WrapperOnManifold
    inline void applyDiff() const;
  public:
    /// \brief Gets the manifold
    virtual const mnf::Manifold* getManifold() const;

    static const flag_t flags = ROBOPTIM_IS_ON_MANIFOLD;

    virtual flag_t getFlags() const;
  };

  template <typename T>
  std::ostream&
  operator<<(std::ostream& o, const WrapperOnManifold<T>& instWrap)
  {
    return instWrap.print(o);
  }

  /// @}

} // end of namespace roboptim.


# include <roboptim/core/manifold-map/decorator/wrapper-on-manifold.hxx>
#endif //! ROBOPTIM_CORE_MANIFOLD_MAP_DECORATOR_WRAPPER_ON_MANIFOLD_HH
