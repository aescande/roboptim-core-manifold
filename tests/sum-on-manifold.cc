// Copyright (C) 2015 by Benjamin Chrétien, CNRS-UM LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>
#include <memory>

#include <roboptim/core/differentiable-function.hh>

#include <roboptim/core/manifold-map/decorator/manifold-map.hh>
#include <roboptim/core/manifold-map/decorator/sum-on-manifold.hh>
#include <roboptim/core/manifold-map/decorator/wrapper-on-manifold.hh>

#include <manifolds/RealSpace.h>

typedef boost::mpl::list< ::roboptim::EigenMatrixDense,
			  ::roboptim::EigenMatrixSparse> functionTypes_t;

template<class T>
struct F : public roboptim::GenericDifferentiableFunction<T>
{
  ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
  (roboptim::GenericDifferentiableFunction<T>);

  F (size_type n, value_type k)
    : roboptim::GenericDifferentiableFunction<T> (n, 1, "f (x) = k Σ(x_i)"),
      k_ (k)
  {}

  void impl_compute (result_ref res, const_argument_ref x) const
  {
    res[0] = k_ * x.sum ();
  }

  void impl_gradient (gradient_ref grad, const_argument_ref, size_type) const
  {
    grad.setConstant (k_);
  }

private:
  value_type k_;
};

template <>
inline void
F<roboptim::EigenMatrixSparse>::impl_gradient(gradient_ref grad,
                                              const_argument_ref,
                                              size_type) const
{
  grad.setZero ();
  for (size_type i = 0; i < inputSize (); ++i)
    {
      grad.insert (i) = k_;
    }
}

boost::shared_ptr<boost::test_tools::output_test_stream> output;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE_TEMPLATE (sum_on_manifold_test_0, T, functionTypes_t)
{
  output = retrievePattern("sum-on-manifold");

  typedef F<T> testF_t;
  typedef typename testF_t::value_type value_type;

  // Create dummy functions
  static const int n = 5;
  const value_type k0 = 2.;
  const value_type k1 = 3.;
  std::shared_ptr<testF_t> f0 = std::make_shared<testF_t> (n, k0);
  std::shared_ptr<testF_t> f1 = std::make_shared<testF_t> (n, k1);

  (*output) << *f0 << std::endl;
  (*output) << *f1 << std::endl;

  // Define manifolds
  mnf::RealSpace fManif (n);
  fManif.name () = "F manifold";
  mnf::RealSpace pbManif (1);
  pbManif.name () = "Problem manifold";

  ROBOPTIM_DESC_MANIFOLD(fManif_t, ROBOPTIM_REAL_SPACE(n));
  ROBOPTIM_NAMED_FUNCTION_BINDING(Desc_F_Manif, testF_t, fManif_t);

  Desc_F_Manif descWrapF0 (n, k0);
  Desc_F_Manif descWrapF1 (n, k1);

  mnf::RealSpace dummy (n);

  (*output) << descWrapF0 << std::endl;
  (*output) << descWrapF1 << std::endl;

  // Create Adder helper and add the functions
  typedef roboptim::AdderOnManifold<T> adder_t;
  typedef F<T> testF_t;
  adder_t adder;
  typename adder_t::functionPtr_t fSum;
  BOOST_CHECK_THROW (fSum = adder.getFunction (pbManif), std::runtime_error);
  adder.add (0.2, descWrapF0, dummy);
  adder.add (0.8, descWrapF1, dummy);

  // Get the sum of these functions on manifolds
  fSum = adder.getFunction (pbManif);
  (*output) << *fSum << std::endl;

  std::cout << output->str() << std::endl;
  BOOST_CHECK (output->match_pattern());
}

BOOST_AUTO_TEST_SUITE_END ()
