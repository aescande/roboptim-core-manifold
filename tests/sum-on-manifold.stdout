f (x) = k Σ(x_i) (differentiable function)
f (x) = k Σ(x_i) (differentiable function)
DescriptiveWrapper's function:
  Name: f (x) = k Σ(x_i)
  Input size: 5
  Output size: 1
DescriptiveWrapper's manifold:
  Name: R5
  Elementary: yes
  Dimension: 5
DescriptiveWrapper's function:
  Name: f (x) = k Σ(x_i)
  Input size: 5
  Output size: 1
DescriptiveWrapper's manifold:
  Name: R5
  Elementary: yes
  Dimension: 5
Wrapper on manifold:
  Function name: Sum("f (x) = k Σ(x_i)", "f (x) = k Σ(x_i)")
  Restricted mapping: 0, 0, 0, 0, 0
